gobblin-runtime/src/main/java/org/apache/gobblin/runtime/mapreduce/DistributedClasspathManager.java

```diff
@@ -0,0 +1,234 @@
+
+    if (!this.fs.exists(this.rootPath)) {
+      this.fs.mkdirs(this.rootPath);
+      this.fs.setPermission(this.rootPath, new FsPermission(FsAction.ALL, FsAction.ALL, FsAction.ALL));
+    }
+  }
+
+  /**
+   * Add the artifact containing the specified class to the classpath of the input job.
+   */
+  public void addToDistributedClasspathByClass(Job job, Class<?> klazz, Path privateJarsDir, int maxAttempts)
+      throws IOException {
+
+    String jarStr = ClassUtil.findContainingJar(klazz);
+    if (Strings.isNullOrEmpty(jarStr)) {
+      // Usually happens when running on IDE
+      return;
+    }
+    if (fs.getScheme().equals("file")) {
+      // not running in distributed mode
+      return;
+    }
+    addFileToClasspath(job, lfs.getFileStatus(new Path(jarStr)), privateJarsDir, maxAttempts);
+  }
+
+  /**
+   * Add the artifacts contained in the {@code jarFileList} to the classpath of the input job.
+   */
+  public void addToDistributedClasspath(Job job, String jarFileList, Path privateJarsDir, int maxAttempts)
+      throws IOException {
+
+    if (Strings.isNullOrEmpty(jarFileList)) {
+      return;
+    }
+    if (fs.getScheme().equals("file")) {
+      // not running in distributed mode
+      return;
+    }
+
+    // jarFileList is a ',' separated string
+    for (String jarFile : SPLITTER.split(jarFileList)) {
+      Path srcJarFile = new Path(jarFile);
+      // A jar file can be specified with glob pattern
+      for (FileStatus status : lfs.globStatus(srcJarFile)) {
+        addFileToClasspath(job, status, privateJarsDir, maxAttempts);
+      }
+    }
+  }
+
+  private void addFileToClasspath(Job job, FileStatus fileStatus, Path privateJarsDir, int maxAttempts)
+      throws IOException {
+
+    Path file = fileStatus.getPath();
+
+    // SNAPSHOT files should not be shared, as different jobs may be using different versions of it
+    Path baseDir = file.getName().contains("SNAPSHOT") ? privateJarsDir : this.rootPath;
+    // DistributedCache requires absolute path, so we need to use makeQualified.
+    Path destFilePath = new Path(this.fs.makeQualified(baseDir), file.getName());
+
+    // For each FileStatus there are chances it could fail in copying at the first attempt, due to file-existence
+    // or file-copy is ongoing by other job instance since all Gobblin jobs share the same jar file directory.
+    int attempt = 1;
+    while (true) {
+      try {
+        if (this.fs.exists(destFilePath)) {
+          // Assuring destJarFile exists and the size of file on targetPath matches the one on local path.
+          if (this.fs.getFileStatus(destFilePath).getLen() == fileStatus.getLen()) {
+            break;
+          }
+          Thread.sleep(WAITING_TIME_ON_IMCOMPLETE_UPLOAD);
+        } else {
+          // Copy the file from local file system to HDFS
+          this.fs.copyFromLocalFile(file, destFilePath);
+          this.fs.setPermission(destFilePath, new FsPermission(FsAction.ALL, FsAction.ALL, FsAction.ALL));
```

Comment:
is that 777? better to be more restrictive by default

Reply:
Yeah. I can make it to 750.